### PR TITLE
Refactor Coalesce and Normalizable.

### PR DESCRIPTION
**Pull request by sellout**
This currently fails to compile with the following errors:
```
[error] /Users/greg/Documents/SlamData/quasar/connector/src/main/scala/quasar/qscript/Coalesce.scala:205: can't existentially abstract over parameterized type F
[error]     new Coalesce[QScriptCore] {
[error]     ^
[error] /Users/greg/Documents/SlamData/quasar/connector/src/main/scala/quasar/qscript/Coalesce.scala:362: can't existentially abstract over parameterized type F
[error]     new Coalesce[ProjectBucket] {
[error]     ^
[error] /Users/greg/Documents/SlamData/quasar/connector/src/main/scala/quasar/qscript/Coalesce.scala:398: can't existentially abstract over parameterized type F
[error]     new Coalesce[ThetaJoin] {
[error]     ^
[error] /Users/greg/Documents/SlamData/quasar/connector/src/main/scala/quasar/qscript/Coalesce.scala:436: can't existentially abstract over parameterized type F
[error]     new Coalesce[EquiJoin] {
[error]     ^
[error] four errors found
```
I found http://stackoverflow.com/questions/3122398/cant-existentially-abstract-over-parameterized-type, but ¯\\\_(ツ)_/¯

### CODE REVIEW

**Comment by sellout** [comment]
Ok, got this compiling (thanks to @paulp), and updated. In review now.

File: connector/src/main/scala/quasar/qscript/Normalizable.scala [hunk 1]
```diff
@@ -29,13 +34,170 @@ import simulacrum.typeclass
...
+            tj.f,
+            combineNorm.getOrElse(tj.combine)).some
+      }))
+
+  def QScriptCore = {
+    // NB: all single-bucket reductions should reduce on `null`
+    def normalizeBucket(bucket: FreeMap): FreeMap = bucket.resume.fold({
+      case MapFuncs.Constant(_) => MapFuncs.NullLit[T, Hole]()
+      case _                    => bucket
+    }, κ(bucket))
```
Comment (paulp) [inline connector/src/main/scala/quasar/qscript/Normalizable.scala]: `normalizeBucket` is never used.

File: connector/src/main/scala/quasar/qscript/Normalizable.scala [hunk 2]
```diff
@@ -29,13 +34,170 @@ import simulacrum.typeclass
...
+
+    make(λ[QScriptCore ~> (Option ∘ QScriptCore)#λ] {
+      case Reduce(src, bucket, reducers, repair) => {
+        val reducersOpt: List[Option[ReduceFunc[FreeMap]]] =
+          reducers.map(_.traverse(freeMFEq[Hole](_)))
+
+        val reducersNormOpt: Option[List[ReduceFunc[FreeMap]]] =
+          (!reducersOpt.map(_.toList).flatten.isEmpty).option(
```
Comment (paulp) [inline connector/src/main/scala/quasar/qscript/Normalizable.scala]: The outer parens aren't doing anything.

File: connector/src/main/scala/quasar/qscript/Normalizable.scala [hunk 3]
```diff
@@ -29,13 +34,170 @@ import simulacrum.typeclass
...
+        val orderOpt: List[Option[(FreeMap, SortDir)]] =
+          order.map {
+            _.leftMap(freeMFEq(_)) match {
+              case (Some(fm), dir) => Some((fm, dir))
+              case (_, _)          => None
+            }
+          }
+
+        val orderNormOpt: Option[List[(FreeMap, SortDir)]] =
+          (!orderOpt.map(_.toList).flatten.isEmpty).option(
```
Comment (paulp) [inline connector/src/main/scala/quasar/qscript/Normalizable.scala]: Same with these parens. Also both expressions are super indirect, it looks like they could both be `x.exists(_.nonEmpty)`.

File: connector/src/main/scala/quasar/qscript/Normalizable.scala [hunk 4]
```diff
@@ -29,13 +34,170 @@ import simulacrum.typeclass
 
 // it would be nice to use the `NTComp` alias here, but it cannot compile
 trait NormalizableInstances {
+  def normalizable[T[_[_]] : Recursive : Corecursive : EqualT : ShowT] =
+    new NormalizableT[T]
+
   implicit def const[A] = new Normalizable[Const[A, ?]] {
     def normalizeF = λ[Const[A, ?] ~> (Option ∘ Const[A, ?])#λ](_ => None)
   }
```
Comment (paulp) [inline connector/src/main/scala/quasar/qscript/Normalizable.scala]: This one could be 
```scala
  implicit def const[A] =
    make(λ[Const[A, ?] ~> (Option ∘ Const[A, ?])#λ](_ => None))
```

**Comment by sellout** [comment]
Updated.

**Pull request merged**